src/CoreWCF.Primitives/src/CoreWCF/Channels/MetadataMiddleware.cs

```diff
@@ -0,0 +1,245 @@
+                    }
+                    else
+                    {
+                        httpMetadataUri = fallbackHttpEndpointAddress;
+                    }
+
+                    if(httpMetadataUri != null)
+                    {
+                        MapMetadata(branchApp, httpMetadataUri.AbsolutePath, metadataExtension.ConfigureWith(httpMetadataUri, false));
+                    }
+                }
+
+                if (metadataExtension.HttpHelpPageEnabled)
+                {
+                    Uri helpPageUri;
+                    if (metadataExtension.HttpHelpPageUrl != null && metadataExtension.HttpHelpPageUrl.AbsolutePath != "/")
+                    {
+                        helpPageUri = metadataExtension.HttpHelpPageUrl;
+                    }
+                    else
+                    {
+                        helpPageUri = fallbackHttpEndpointAddress;
+                    }
+                    if (helpPageUri != null && !helpPageUri.Equals(httpMetadataUri))
+                    {
+                        // Only map the help page uri if it's not null, and different from http metadata uri.
+                        MapMetadata(branchApp, helpPageUri.AbsolutePath, metadataExtension.ConfigureWith(helpPageUri, false));
+                    }
+                }
+
+                if (metadataExtension.HttpsGetEnabled)
+                {
+                    if (metadataExtension.HttpsGetUrl != null && metadataExtension.HttpsGetUrl.AbsolutePath != "/")
+                    {
+                        httpsMetadataUri = metadataExtension.HttpsGetUrl;
+                    }
+                    else
+                    {
+                        httpsMetadataUri = fallbackHttpsEndpointAddress;
+                    }
+
+                    if (httpsMetadataUri != null)
+                    {
+                        MapMetadata(branchApp, httpsMetadataUri.AbsolutePath, metadataExtension.ConfigureWith(httpsMetadataUri, true));
+                    }
+                }
+
+                if (metadataExtension.HttpsHelpPageEnabled)
+                {
+                    Uri helpPageUri;
+                    if (metadataExtension.HttpsHelpPageUrl != null && metadataExtension.HttpsHelpPageUrl.AbsolutePath != "/")
+                    {
+                        helpPageUri = metadataExtension.HttpsHelpPageUrl;
+                    }
+                    else
+                    {
+                        helpPageUri = fallbackHttpsEndpointAddress;
+                    }
+                    if (helpPageUri != null && !helpPageUri.Equals(httpsMetadataUri))
+                    {
+                        // Only map the help page uri if it's not null, and different from https metadata uri.
+                        MapMetadata(branchApp, helpPageUri.AbsolutePath, metadataExtension.ConfigureWith(helpPageUri, true));
+                    }
+                }
+            }
+
+            branchApp.Use(_ => {
```

Comment:
I've reworked this, but we need potentially 4 mappings. HTTP and HTTPS for the help page and the wsdl. There's code to avoid duplicate mapping when the help page and wsdl are at the same url, but we do still need one for each of HTTP and HTTPS is they are both enabled.